File: web/src/main/java/top/bettercode/summer/web/resolver/KilogramToGramConverter.java
package top.bettercode.summer.web.resolver;

import java.util.Collections;
import java.util.Set;
import org.jetbrains.annotations.NotNull;
import org.springframework.core.convert.TypeDescriptor;
import org.springframework.core.convert.converter.ConditionalGenericConverter;
import org.springframework.util.StringUtils;
import top.bettercode.summer.web.support.KilogramUtil;

public class KilogramToGramConverter implements ConditionalGenericConverter {

  @Override
  public boolean matches(@NotNull TypeDescriptor typeDescriptor, TypeDescriptor typeDescriptor1) {
    return typeDescriptor1.hasAnnotation(KilogramToGram.class);
  }

  @Override
  public Set<ConvertiblePair> getConvertibleTypes() {
    return Collections.singleton(new ConvertiblePair(String.class, Long.class));
  }

  @Override
  public Object convert(Object object, @NotNull TypeDescriptor sourceType,
      @NotNull TypeDescriptor targetType) {
    if (!StringUtils.hasText((String) object)) {
      return null;
    }
    return KilogramUtil.toGram((String) object);
  }
}

File: web/src/main/kotlin/top/bettercode/summer/web/resolver/KilogramToGramConverter.kt
package top.bettercode.summer.web.resolver

import org.springframework.core.convert.TypeDescriptor
import org.springframework.core.convert.converter.ConditionalGenericConverter
import org.springframework.core.convert.converter.GenericConverter
import org.springframework.util.StringUtils
import top.bettercode.summer.web.support.KilogramUtil.toGram

class KilogramToGramConverter : ConditionalGenericConverter {
    override fun matches(typeDescriptor: TypeDescriptor, typeDescriptor1: TypeDescriptor): Boolean {
        return typeDescriptor1.hasAnnotation(KilogramToGram::class.java)
    }

    override fun getConvertibleTypes(): Set<GenericConverter.ConvertiblePair> {
        return setOf(GenericConverter.ConvertiblePair(String::class.java, Long::class.java))
    }

    override fun convert(`object`: Any?, sourceType: TypeDescriptor, targetType: TypeDescriptor): Any? {
        return if (!StringUtils.hasText(`object` as String)) {
            null
        } else toGram(`object`)
    }
}
